command/agent/config/config_test.go

```diff
@@ -116,8 +116,236 @@ func TestLoadConfigFile_AgentCache(t *testing.T) {
+					TLSDisable: true,
+				},
+				{
+					Type:        "tcp",
+					Role:        "default",
+					Address:     "127.0.0.1:8400",
+					TLSKeyFile:  "/path/to/cakey.pem",
+					TLSCertFile: "/path/to/cacert.pem",
+				},
+			},
+		},
+		AutoAuth: &AutoAuth{
+			Method: &Method{
+				Type:      "aws",
+				MountPath: "auth/aws",
+				Config: map[string]interface{}{
+					"role": "foobar",
+				},
+			},
+			Sinks: []*Sink{
+				{
+					Type:   "file",
+					DHType: "curve25519",
+					DHPath: "/tmp/file-foo-dhpath",
+					AAD:    "foobar",
+					Config: map[string]interface{}{
+						"path": "/tmp/file-foo",
+					},
+				},
+			},
+		},
+		APIProxy: &APIProxy{
+			UseAutoAuthToken:   true,
+			ForceAutoAuthToken: false,
+		},
+		Cache: &Cache{
+			UseAutoAuthToken:    true,
+			UseAutoAuthTokenRaw: true,
+			ForceAutoAuthToken:  false,
+			Persist: &Persist{
+				Type:                    "kubernetes",
+				Path:                    "/vault/agent-cache/",
+				KeepAfterImport:         true,
+				ExitOnErr:               true,
+				ServiceAccountTokenFile: "/tmp/serviceaccount/token",
+			},
+		},
+		Vault: &Vault{
+			Address:          "http://127.0.0.1:1111",
+			CACert:           "config_ca_cert",
+			CAPath:           "config_ca_path",
+			TLSSkipVerifyRaw: interface{}("true"),
+			TLSSkipVerify:    true,
+			ClientCert:       "config_client_cert",
+			ClientKey:        "config_client_key",
+			Retry: &Retry{
+				NumRetries: 12,
+			},
+		},
+	}
+
+	config.Prune()
+	if diff := deep.Equal(config, expected); diff != nil {
+		t.Fatal(diff)
+	}
+
+	config, err = LoadConfigFile("./test-fixtures/config-dir-cache/config-cache1.hcl")
+	if err != nil {
+		t.Fatal(err)
+	}
+	config2, err := LoadConfigFile("./test-fixtures/config-dir-cache/config-cache2.hcl")
+
+	mergedConfig := config.Merge(config2)
+
+	mergedConfig.Prune()
+	if diff := deep.Equal(mergedConfig, expected); diff != nil {
+		t.Fatal(diff)
+	}
+}
+
+func TestLoadConfigDir_AutoAuthAndListener(t *testing.T) {
+	config, err := LoadConfig("./test-fixtures/config-dir-auto-auth-and-listener/")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected := &Config{
+		SharedConfig: &configutil.SharedConfig{
+			PidFile: "./pidfile",
+			Listeners: []*configutil.Listener{
+				{
+					Type:       "tcp",
+					Address:    "127.0.0.1:8300",
+					TLSDisable: true,
+				},
+			},
+		},
+		AutoAuth: &AutoAuth{
+			Method: &Method{
+				Type:      "aws",
+				MountPath: "auth/aws",
+				Config: map[string]interface{}{
+					"role": "foobar",
+				},
+			},
+			Sinks: []*Sink{
+				{
+					Type:   "file",
+					DHType: "curve25519",
+					DHPath: "/tmp/file-foo-dhpath",
+					AAD:    "foobar",
+					Config: map[string]interface{}{
+						"path": "/tmp/file-foo",
+					},
+				},
+			},
+		},
+	}
+
+	config.Prune()
+	if diff := deep.Equal(config, expected); diff != nil {
+		t.Fatal(diff)
+	}
+
+	config, err = LoadConfigFile("./test-fixtures/config-dir-auto-auth-and-listener/config1.hcl")
```

Comment:
Why these extra checks?  They seem redundant after the deep.Equal above.

Reply:
Essentially, we are doing two checks:

1. Check that the config matches if we load the directory
2. Check that the config matches if we load and merge the two files independently

In essence, we're confirming there is no difference between:
`vault agent -config=dir1`
and
`vault agent -config=dir1/file1.hcl -config=dir1/file2.hcl`

Reply:
I guess I'm always dubious of the value of tests that reproduce the code they're testing.

Reply:
I think I'd like to keep it as is, to ensure that the behaviour doesn't diverge in the future. Anecdotally, this was quite useful during development. We could factor out the test to two tests that check the multi-file approach in a different test to the dir approach, but we'd lose the validation that the two produce the same resultant config.